fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/model/builder/ModelBuilder.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.datagen.v1.model.builder;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import net.minecraft.data.client.ItemModelGenerator;
+import net.minecraft.data.client.Model;
+import net.minecraft.data.client.TextureKey;
+import net.minecraft.data.client.TextureMap;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.datagen.v1.model.property.DisplayBuilder;
+import net.fabricmc.fabric.api.datagen.v1.model.property.ElementBuilder;
+
+/**
+ * Base builder class used for general model files. Can be extended if necessary to allow for any custom models not typically associated with blocks and items.
+ *
+ * <p>Note that custom non-block/item models may generally need to be generated under {@link net.fabricmc.fabric.api.datagen.v1.provider.FabricModelProvider#generateItemModels(ItemModelGenerator)}.
+ */
+@SuppressWarnings("unchecked")
+public abstract class ModelBuilder<T extends ModelBuilder<T>> {
+	protected final Identifier parent;
+	protected final Set<TextureKey> requiredTextures = new HashSet<>();
+	protected final HashMap<TextureKey, Identifier> textures = new HashMap<>();
+	protected final EnumMap<DisplayBuilder.Position, DisplayBuilder> displays = new EnumMap<>(DisplayBuilder.Position.class);
+	protected final List<ElementBuilder> elements = new ArrayList<>();
```

Comment:
Should be priviate. Either add a protected setter or use the existing builder methods when copying.